File: src/Problem014.go
package main

import (
	"fmt"
	"time"
)

const top = 1000000

var memo map[uint]uint

func collatz(n uint) uint {
	if ans, ok := memo[n]; ok {
		return ans
	}

	if n%2 == 0 {
		memo[n] = collatz(n/2) + 1
		return collatz(n)
	}

	memo[n] = collatz(3*n+1) + 1
	return collatz(n)

}

func main() {
	starttime := time.Now()

	memo = make(map[uint]uint)
	memo[1] = 1
	var best, where uint

	for i := uint(1); i <= top; i++ {
		if collatz(i) > best {
			best = collatz(i)
			where = i
			// fmt.Println(i, best)
		}
	}

	fmt.Println(where)

	fmt.Println("Elapsed time:", time.Since(starttime))
}

File: python/Problem014.py
collatz = dict({1:1})

def chainlength(n):
    if n in collatz:
        return collatz[n]
    if n%2 ==0:
        collatz[n] = 1 + chainlength(n/2)
        return chainlength(n)
    collatz[n] =  1 + chainlength(3*n +1)
    return chainlength(n)

best, size = -1, -1

for x in xrange(1,1000000):
    if chainlength(x) > size:
        # print "{}\t{}".format(x, chainlength(x))
        best, size =x, chainlength(x)

print best